modules/core/src/main/java/org/apache/ignite/internal/client/thin/ReliableChannel.java

```diff
@@ -461,9 +418,8 @@ private void initAllChannelsAsync() {
      * @param ch Channel.
      */
     private void onTopologyChanged(ClientChannel ch) {
-        if (partitionAwarenessEnabled && affinityCtx.updateLastTopologyVersion(ch.serverTopologyVersion(),
-            ch.serverNodeId()))
-            initAllChannelsAsync();
+        if (affinityCtx.updateLastTopologyVersion(ch.serverTopologyVersion(), ch.serverNodeId()))
+            channelsInit(true);
```

Comment:
> we can do it sync in sending thread on channel failure and if we detect topology change before

It can be wrong. Topology change can be triggered by joining node, so channel failure may never occur. I'll make async initialization in any case.

Reply:
If partition awareness is disabled and a new node is joined it's fine to work with the current node until failure (we don't need to find new nodes unless we disconnected from the current).

Reply:
Fair. Fixed it. Change signature of `initChannelHolders`, now it's not void but returns boolean (flag of interruption). Otherwise for disabled awareness and continuous topology events there would be a chances for StackOverflow, as channel failure trigger recursively `applyOnDefaultChannel` to enable default channel.